### ISSUE TITLE
Consider changing socket close behavior to graceful shutdown when outstanding IO is cancelled

### ISSUE THREAD

**Issue by geoffkizer** [dotnet/runtime#43385]
When you Close a Socket, we cancel any pending IO on that Socket. Currently (and going back to .NET Framework days), if any pending IO operations are actually cancelled, the close changes from a graceful shutdown (FIN) to an abortive shutdown (RST). 

I am not sure why this behavior exists, but I think it is problematic and we should consider changing it so that we still perform a graceful shutdown in these cases.

It's fairly common for users to write a receive loop that continually pulls data from a socket. We do this in our HTTP2 implementation. Because of the above behavior, HTTP2 connections are typically going to be terminated with a RST instead of a FIN, which is not ideal.

What's worse is that this behavior isn't deterministic in most cases in practice. Consider the receive loop from above -- most of the time it is waiting for data, and thus has a pending receive IO. But if you happen to call Close when the receive loop has completed a previous IO and hasn't yet issued the next IO, then you will not get an abortive close in this case. Similarly for the send side.

Users can force an abortive shutdown (RST) by calling Close(0) on the socket. If they don't do this, then presumably they want a graceful shutdown, and we should not override this, especially since we can't do it in a predictable/deterministic way.

@antonfirsov @tmds @wfurt @stephentoub

**Comment by scalablecory** [comment]
Do we have any composite operations in sockets right now? We would if we updated `SendFile` to transfer in 2GB increments.

A graceful shutdown mid-op would be weird.

**Comment by scalablecory** [comment]
Actually, I think for the same reason you'd want cancel to be graceful during a read loop, you'd want it to be abortive in a write loop. I don't know if there is a correct answer here...

**Comment by geoffkizer** [comment]
Yeah, unfortunately, we cannot distinguish between pending reads and pending writes.

That said, the same timing issue exists with a write loop: if you happen to call Close while there is not actually an outstanding write, you won't get an abortive close.

I think the ultimate answer is that the user needs to decide on graceful vs abortive themselves, based on the state of their app or whatever. The problem is, we're not letting them do this -- we are changing our behavior based on whether there's an outstanding IO at the moment or not.

**Comment by scalablecory** [comment]
I think `Close()`/`Dispose()` should be abortive if you haven't called `Shutdown()` -- very weird that you don't get an abortive close if there are no pending I/Os, are you sure?

**Comment by geoffkizer** [comment]
> I think Close()/Dispose() should be abortive if you haven't called Shutdown() -- very weird that you don't get an abortive close if there are no pending I/Os, are you sure? Sounds like a bug.

I just tested it to confirm, and it is graceful if there's no pending IOs.

That said, if we did make Close/Dispose abortive if you haven't called Shutdown(), then at least that would be consistent and deterministic.

**Comment by geoffkizer** [comment]
FWIW, NetworkStream calls Shutdown in its Dispose (actually InternalShutdown(), but the effect should be the same).

**Comment by scalablecory** [comment]
> I think `Close()`/`Dispose()` should be abortive if you haven't called `Shutdown()` -- very weird that you don't get an abortive close if there are no pending I/Os, are you sure? Sounds like a bug.

Ugh, I forgot about linger. Yeah, we shouldn't do this (my suggestion) -- it is different from how normal BSD sockets work.

**Comment by wfurt** [comment]
This is what Unix generally does e.g. close() would flush gracefully all pending data. All the new weird code was added to mimic Windows. (and one would need special call if Reset is really desired)
I think this would be breaking change but I would personally see it as improvement.

**Comment by antonfirsov** [comment]
@scalablecory we also have some reliability, platform consistency, and code complexity problems around Dispose (see Geoff's email).

It looks like we can overcome these issues easily by altering the default behavior to be graceful.

**Comment by tmds** [comment]
> What's worse is that this behavior isn't deterministic in most cases in practice. 

Because it is nondeterministic, I also think it makes sense to make Close perform a FIN close even when there are out-standing operations.

When the Socket gets closed by the finalizer, maybe we still want a RST close.

**Comment by geoffkizer** [comment]
> When the Socket gets closed by the finalizer, maybe we still want a RST close.

Agreed, this seems like a case where RST definitely is the appropriate action.
